### ISSUE TITLE
[CRASH] Crash in tcpconn_add

### ISSUE THREAD

**Issue by bcnewlin** [OpenSIPS/opensips#2344]
**OpenSIPS version you are running**

```
# opensips -V
version: opensips 2.4.8 (x86_64/linux)
flags: STATS: On, DISABLE_NAGLE, SHM_MMAP, PKG_MALLOC, F_MALLOC, FAST_LOCK-ADAPTIVE_WAIT
ADAPTIVE_WAIT_LOOPS=1024, MAX_RECV_BUFFER_SIZE 262144, MAX_LISTEN 16, MAX_URI_SIZE 1024, BUF_SIZE 65535
poll method support: poll, epoll, sigio_rt, select.
git revision: a88a57479
main.c compiled on 01:39:42 Dec  5 2020 with gcc 7
```

**Crash Core Dump**

https://pastebin.com/fhbhskAh

**Describe the traffic that generated the bug**

Normal production traffic, unable to isolate a specific cause

**To Reproduce**

Unknown

**Relevant System Logs**

```
Dec 15 22:45:15 [7877] INFO:core:probe_max_sock_buff: using snd buffer of 416 kb
Dec 15 22:45:15 [7877] INFO:core:init_sock_keepalive: TCP keepalive enabled on socket 706
Dec 15 22:45:15 [7877] CRITICAL:core:sig_usr: segfault in process pid: 7877, id: 33
Dec 15 22:45:15 XXhostnameXX kernel: opensips[7877]: segfault at 0 ip 0000000000414ab6 sp 00007ffe1324bb70 error 6 in opensips[400000+30c000]
```

**OS/environment information**
 - Operating System:  Amazon linux 2
 - OpenSIPS installation:  git
 - other relevant information:

**Additional context**

**Comment by bogdan-iancu** [comment]
@bcnewlin , could you do in frame 2:
```
p *c
```

**Comment by bcnewlin** [comment]
@bogdan-iancu Yes, I will. Can you remind me again how to get to frame 2?

**Comment by bogdan-iancu** [comment]
Jus to `f 2`Sent from my Samsung Galaxy smartphone.
-------- Original message --------From: Ben Newlin <notifications@github.com> Date: 12/17/20  19:57  (GMT+02:00) To: OpenSIPS/opensips <opensips@noreply.github.com> Cc: Bogdan Andrei IANCU <bogdan@opensips.org>, Mention <mention@noreply.github.com> Subject: Re: [OpenSIPS/opensips] [CRASH] Crash in tcpconn_add (#2344) 
@bogdan-iancu Yes, I will. Can you remind me again how to get to frame 2?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or unsubscribe.
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/OpenSIPS/opensips/issues/2344#issuecomment-747600627",
"url": "https://github.com/OpenSIPS/opensips/issues/2344#issuecomment-747600627",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

**Comment by bogdan-iancu** [comment]
Any update here @bcnewlin  ?

**Comment by bcnewlin** [comment]
@bogdan-iancu Apologies, I have not had time to gather the extra information. I will do so as soon as possible.

**Comment by bcnewlin** [comment]
@bogdan-iancu Here you go: https://pastebin.com/HC338n3E

**Comment by bogdan-iancu** [comment]
@bcnewlin , were you running kind of a long time stress test ? It looks like the connection ID did an overflow `id = -2147483647` ( 2^31 = 2147483648), jumping from 2147483647 to -2147483647 . We should move the `connection_id` counter (and the ids) from `int` to `unsigned int`

**Comment by bcnewlin** [comment]
@bogdan-iancu I cannot recall if this issue occurred in our testing or production environment. In our testing environment we do run long time stress tests pretty much constantly. In our production environment OpenSIPS can run at longest for just under a month without restarts.

**Comment by bogdan-iancu** [comment]
I will give it some time on dev/master to see for potential side-effects. If everything ok, I will do a backport.

**Comment by bcnewlin** [comment]
Thanks @bogdan-iancu! We are still on 2.4, would you mind leaving this open and updating once the backport is available? Alternatively if you can provide a patch we can test it out as well pending the backport.

**Comment by bogdan-iancu** [comment]
@bcnewlin , the fix was backported to 3.1 and 2.4. It is already part of 3.2 and head also.

**Comment by bcnewlin** [comment]
@bogdan-iancu Thank you!
